lib/ecto/enum.ex

```diff
@@ -235,6 +244,47 @@ defmodule Ecto.Enum do
     |> Keyword.keys()
   end
 
+  @doc """
+  Returns the possible cast values for a given schema or types map and field.
+
+  These values are all the values that the `cast/2` function takes in.
+
+  ## Examples
+
+  Assuming this schema:
+
+      defmodule MySchema do
+        use Ecto.Schema
+
+        schema "my_schema" do
+          field :my_string_enum, Ecto.Enum, values: [:foo, :bar, :baz]
+          field :my_integer_enum, Ecto.Enum, values: [foo: 1, bar: 2, baz: 5]
+        end
+      end
+
+  Then:
+
+      Ecto.Enum.cast_values(MySchema, :my_string_enum)
+      #=> [:foo, "foo", :bar, "bar", :baz, "baz"]
+
+      Ecto.Enum.cast_values(MySchema, :my_integer_enum)
+      #=> [:foo, "foo", 1, :bar, "bar", 2, :baz, "baz", 5]
+
+  """
+  def cast_values(schema_or_types, field) do
+    schema_or_types
+    |> mappings(field)
+    |> cast_values_from_mappings()
+  end
+
+  defp cast_values_from_mappings(mappings) do
+    mappings
+    |> Stream.flat_map(fn {key, value} ->
+      [key, to_string(key), value]
```

Comment:
I am not sure. The `cast/2` function really works with all those 3 values... so it makes sense that we include all 3. More so if `validation: :inclusion` where the error message can be understood as `"must be one of %{enum}"`.

It is weird if we were to, say, emit this error from a JSON API, where atoms are transformed into strings. So `[:foo, "foo", 1]` would be transformed into `["foo", "foo", 1]` by Jason and then we have duplicate values. But then I think it is on the final user to add a `Enum.uniq/2` call, and not us here in Ecto.

But all in all, I think it is fine that we leave this as is 🤷

Reply:
My concern is that they can be inaccurate and we over-report. Or perhaps the mistake was in including dumped values by default. But I am certain we should not include both cast and load keys. load keys are atoms and it doesn't make sense to expose atoms (an Elixir construct) to an external system (which is allegedly when we call cast).

Reply:
Agreed! However, us for example, we use embedded schemas as a replacement for vanilla structs and as a way to protect ourselves from ourselves.

We define functions like `User.new/1` that rely on changesets under the hood. And in this case it would make sense that we do `User.new(type: :admin)`. But maybe the `:admin` type was at some point removed, so this call will now fail. And because in this case schemas are used from within the system, and not necessarily from an external call, then it perhaps makes sense that we include atoms — so for example `enum: [:regular, "regular", :moderator, "moderator"]`. 🤔

Reply:
@josevalim that was mostly just an example though. I am happy to change this to not include atoms, provided that the main intended usage of casting is from external data, as you mentioned. 🙂

And we still accept atoms but that remains as an implicit implementation detail.

Should we do that? Should we ask for a third opinion? Wdyt?

Reply:
I can kind of see it from two ways:

1. Cast is meant for normalizing/validating external data. In which case it only makes sense to include the `on_cast` keys.
2. Since `on_dump` is being utilized in the cast function, maybe some people are using this as a substitute for `validate_inclusion` as well. It seems a bit out of place but since it's there, I'm not sure if this was intentional. If it was intentional then maybe it makes sense for Ecto to be a bit "smart" in its error reporting. i.e. if the user provided an atom then only show the `on_dump` keys. Otherwise show the `on_cast` keys.

Reply:
After looking at the code more carefully, I believe the `on_load` and `on_dump` stuff in cast is meant for queries not for changesets. So my vote is for just publicizing `on_cast` keys in the error.

Reply:
I am convinced 😄 I have made the necessary changes to only return `on_cast` keys in the error. Thank you both for your inputs! ❤️